terraso_backend/apps/project_management/permission_rules.py

```diff
@@ -12,137 +12,118 @@
 #
 # You should have received a copy of the GNU Affero General Public License
 # along with this program. If not, see https://www.gnu.org/licenses/.
-import rules
-
-from apps.project_management.collaboration_roles import ProjectRole
-
-
-@rules.predicate
-def allowed_to_change_project(user, project):
-    return project.is_manager(user)
-
-
-@rules.predicate
-def allowed_to_add_site_to_project(user, project):
-    return project.is_manager(user) or project.is_contributor(user)
+from dataclasses import dataclass
 
+import rules
 
-@rules.predicate
-def allowed_to_update_site(user, site):
-    if site.owned_by_user:
-        return site.owner == user
-    return site.project.is_manager(user) or site.project.is_contributor(user)
+from apps.project_management.models import Project, Site, SiteNote
 
 
-@rules.predicate
-def allowed_to_delete_site(user, site):
-    if site.owned_by_user:
-        return site.owner == user
-    return site.project.is_manager(user)
+@dataclass
+class Context:
+    # these are the target entity(s) the operation is occurring on. they will always be
+    # referentually consistent (i.e., if the context has a site this will always be the site's
+    # project)
+    project: Project = None
```

Comment:
I realize that we aren't currently using type annotations in our Python codebase (I think we should!), but dataclasses seemed like a good match for this use-case.

Reply:
i agree about type annotations. one thing holding us back is that the django models are our source of truth for pretty much all of our data modeling, but django doesn't ship with type annotations. there is [a community project](https://github.com/typeddjango/django-stubs) but we don't have any experience in this area so it's hard to estimate what the effort:payoff ratio would look like for us